Lib/importlib/metadata.py

```diff
@@ -102,16 +104,24 @@ def extras(self):
         match = self.pattern.match(self.value)
         return list(re.finditer(r'\w+', match.group('extras') or ''))
 
+    @property
+    def distribution(self):
+        if self._distribution is None:
+            self._distribution = Distribution.from_name(self.dist)
+        return self._distribution
+
     @classmethod
-    def _from_config(cls, config):
-        return [
-            cls(name, value, group)
-            for group in config.sections()
-            for name, value in config.items(group)
-            ]
+    def _from_config(cls, dist, config):
+        eps = []
+        for group in config.sections():
+            for name, value in config.items(group):
+                ep = cls(name, value, group, dist.name)
+                ep._distribution = dist
+                eps.append(ep)
+        return eps
```

Comment:
I try to avoid init/for/append loops. Instead:

```suggestion
    def _from_config(cls, dist, config):
        return [
            cls._with_dist(name, value, group, dist)
            for group in config.sections()
            for name, value in config.items(group)
            ]
```

Then in `_with_dist`, implement the logic to construct the object and set the distribution.

Reply:
Sounds reasonable.

Personal question, why? Is using `.append` while iterating the loop slower? Because readability is subjective.

Reply:
It's not about speed but about precision. By creating the list in a single expression, you guarantee in form that the list is constructed uniformly. By using the init/for/append, you introduce the possibility of terminating early or changing some state in the loop, violating safety practices like those encouraged by functional languages. It's more difficult for a reader to ascertain at a glance what the overall behavior is.

Before list/generator expressions were a thing, I recommended instead using `map` and `filter` and similar to achieve the same ends. That code was definitely harder to read and obscure.

I agree the init/for/append is a very common idiom, but it's one I'd like to discourage in favor of the safer, more lisp-like paradigm.